### ISSUE TITLE
Support OAuth 2.0 authentication

### ISSUE THREAD

**Issue by Djaytan** [eclipse-jgit/jgit#94]
### Description

I would like the following code:

```java
String myAccessToken = "access-token-here";
Git git = ...;
git.push()
    .setCredentialsProvider(new OAuthCredentialsProvider(myAccessToken))
    .setTransportConfigCallback(getTransportConfigCallback())
    .setPushAll()
    .call();
```

to include the following HTTP header in the dispatched request toward the Git server:

```
Authorization: Bearer access-token-here
```

### Motivation

Nowadays, OAuth 2.0 is a widespread standard for authorization. This is what is typically used by GitHub and Bitbucket.

However, JGit doesn't support this type of authentication, thus leading to troubles. Fortunately, a workaround exists but ideally it would be better to integrate directly the solution inside JGit.

### Alternatives considered

The workaround which has been found:

```java
String myAccessToken = "access-token-here"
Git git = ...;
git.push()
    .setTransportConfigCallback(
        transport -> {
          if (transport instanceof TransportHttp transportHttp) {
            transportHttp.setAdditionalHeaders(
                Map.of("Authorization", "Bearer " + myAccessToken));
          }
        })
    .setPushAll()
    .call();
```

**Comment by mytest4mail03** [comment]
Im not able to authenticate even with this workaround

**Comment by tomaswolf** [comment]
The linked stack overflow post is about accessing Github repositories via HTTPS and appears to be very outdated and/or confused.

For Github, you need to use a Personal Access Token (PAT), and use that as password in a UsernamePasswordCredentialsProvider. This is described in the [Egit User's Guide](https://github.com/eclipse-egit/egit/wiki/User-Guide#github-tutorial). Using a PAT with Bearer Auth won't work.

**Comment by msohn** [comment]
Support for OAuth 2 authentication is growing amongst git hosting services [1].
Hence I think that's a valid feature request. Actually I just got the same requirement internally at SAP and was about to file a similar feature request :-)

Implementing typical OAuth authentication flows is not as simple as passing a long-term secret like a GitHub personal access token via bearer header to the application. Instead the client's request is redirected to an OAuth server where the user authenticates using long-term credentials (with a validity of months) and potentially a second factor. On successful authentication the server responds with a short-lived token (validity in the minutes) and redirects the client to the application (here github). The client needs to refresh the short-lived tokens before they expire. 

Git has a credentials helper API to retrieve credentials from the user, some credentials store (keychain, cache).
There are implementations of this API supporting OAuth 2, see [2],[3].
I guess we will use some library to implement this client side.

[1] https://github.com/hickford/git-credential-oauth/issues/17
[2] https://github.com/git-ecosystem/git-credential-manager
[3] https://github.com/hickford/git-credential-oauth

**Comment by Djaytan** [comment]
Thanks @msohn for your reply!

I guess some clarifications are required from me in order to better explain what is really expected in this ticket.

### Clarifications about my request

I started from the assumption that an access token was already available when dealing with JGit (typically after the execution of an OAuth 2.0 authorization grant flow). That's why it was more about providing a convenient way to define a bearer authorization header, without considering (yet) further details like a direct support for OAuth 2.0 authorization grant flows. I realize now that it would be better to make the title more specific... Something like "Support Bearer tokens (OAuth 2.0)".

Doing such small change already answers a bunch of needs. Typically, the cases where tokens can be generated right from the Git hosting solution and can be valid for weeks, months or even years (even if frequent rotations are prefered). In such scenarios, applications just read the access token from an environment variable or a file typically.

[Here](https://confluence.atlassian.com/bitbucketserver/http-access-tokens-939515499.html) is an example of process to follow with Bitbucket to generate and then use access tokens. Once generated, the provided usage example that matter to us is the following one:

```
git clone -c http.extraHeader='Authorization: Bearer MDM0MjM5NDc2MDxxxxxxxxxxxxxxxxxxxxx' https://example.com/scm/projectname/teamsinspace.git
```

Since the access token needs to be passed as a bearer token in the authorization header, then we need a convenient way through JGit to configure this kind of credentials.

As of today, only the `UsernamePasswordCredentialsProvider` class is really available to us (ignoring the `NetRCCredentialsProvider` one). However, it does not support credentials already base64-encoded like [the GitHub PATs](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) nor the bearer tokens.

Therefore, the real request here is to provide a convenient way to define [basic](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#basic) and [bearer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#bearer) tokens.

I already provided the alternative actually available to us for bearer tokens. However, I can provide the one for GitHub PATs as well since I saw multiple people being interested by such facility as well. The following should work:

```java
String myGitHubPat = "token-here"
Git git = ...;
git.push()
    .setTransportConfigCallback(
        transport -> {
          if (transport instanceof TransportHttp transportHttp) {
            transportHttp.setAdditionalHeaders(
                Map.of("Authorization", "Basic " + myGitHubPat));
          }
        })
    .setPushAll()
    .call();
```

I think this solution is a bit cleaner than [the one provided on StackOverflow](https://stackoverflow.com/a/28105276/10165346) which consists on exploiting a workaround when dealing with the `UsernamePasswordCredentialsProvider` class:

```java
String remoteUrl = "https://${token}@github.com/user/repo.git";
CredentialsProvider credentialsProvider = new UsernamePasswordCredentialsProvider("${token}", "");
git.push().setRemote(remoteUrl).setCredentialsProvider(credentialsProvider).call();
```

In the end, the target should be (IMHO):

```java
String myBearerToken = "bearer-token-here";
Git git = ...;
git.push()
    .setCredentialsProvider(new HttpBearerTokenCredentialsProvider(myAccessToken))
    .setPushAll()
    .call();
```

and

```java
String myBasicToken = "basic-token-here";
Git git = ...;
git.push()
    .setCredentialsProvider(new HttpBasicTokenCredentialsProvider(myAccessToken))
    .setPushAll()
    .call();
```

With `UsernamePasswordCredentialsProvider` now being a subclass of `HttpBasicTokenCredentialsProvider`, maybe.

_At the beginning I proposed `OAuthCredentialsProvider`, but now I have a preference for `HttpBearerTokenCredentialsProvider` in order to align with `HttpBasicTokenCredentialsProvider` where the prefix correspond to [the HTTP authentication scheme](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes)._

This is why I'm going to rename the issue "Support Bearer tokens (OAuth 2.0) and Basic ones".

### Support OAuth 2.0 authorization grant flows

All of these clarifications now shared about my initial request, it's completely true that having a comprehensive support for OAuth 2.0 makes sense and would be valuable too. That's even more true when we see solutions like GitHub and Bitbucket providing support for some flows (more info [here](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps) and [there](https://developer.atlassian.com/cloud/bitbucket/rest/intro/#oauth-2-0) for the respective Git host platforms). 

In fact, in Amadeus as well we have such needs for administration simplification purposes. Typically, we would like to support interactive flows in order to be able to commit and push changes on behalf of users.

A great source of documentation about the different OAuth 2.0 authorization grant types can be found [here](https://oauth.net/2/grant-types/). A typical interactive flow is [the authorization code with PKCE one](https://oauth.net/2/pkce/).

But I think it would be better to create a separate issue for answering this advanced need.

**Comment by Djaytan** [comment]
> The linked stack overflow post is about accessing Github repositories via HTTPS and appears to be very outdated and/or confused.
> 
> For Github, you need to use a Personal Access Token (PAT), and use that as password in a UsernamePasswordCredentialsProvider. This is described in the [Egit User's Guide](https://github.com/eclipse-egit/egit/wiki/User-Guide#github-tutorial). Using a PAT with Bearer Auth won't work.

You're right @tomaswolf. I updated [my answer](https://stackoverflow.com/a/79051324/10165346) accordingly in order to be as comprehensive as possible in regards of the different setups which may exist. Even if the question is old (created 10 years ago), we still face troubles today, hence this feature request.

**Comment by matthiasblomme** [comment]
Any change in this issue? Now with Azure Repos are moving away from PATs in favour of OAuth authentication this becomes a necessary change

https://learn.microsoft.com/en-us/azure/devops/repos/git/set-up-credential-managers?view=azure-devops

It should also supply the OAuth popup, like the Git CLI offers when credentials are not valid anymore.

**Comment by jetztgradnet** [comment]
We are using JGit in our software as well and would be interested in seeing support for bearer tokens in JGit.

**Comment by kwin** [comment]
See https://github.com/eclipse-egit/egit/issues/191 for an alternative approach.
